Changed GraphSAGE and HinSAGE to take the base generator object

adocherty opened the pull request:
I have updated the classes GraphSAGE and HinSAGE to accept a generator object not a sequence object. This makes them the same as GCN and other full-batch methods. In order to do this the HinSAGE generators now need to have the head node types specified.

Changes:
 * GraphSAGE/HinSAGE models require multiplicity argument if generator not passed
 * HinSAGENodeGenerator requires a head_node_type argument
 * HinSAGELinkGenerator requires a head_node_types argument

### Review Checklist
- [ ] Produced code for required functionality
- [ ] Branch and Pull Request build on CI
- [ ] Branch and Pull Request pass unit tests on CI
- [ ] Docstrings well written
- [ ] Code well commented
- [ ] Changelog updated
- [ ] Checked API documentation
- [ ] Version number reflects new status

geoffj-d61 commented on stellargraph/layer/graphsage.py:
```diff
@@ -1048,56 +1086,70 @@ class DirectedGraphSAGE(GraphSAGE):
         """
-        self.in_samples = generator.generator.in_samples
+
+        if not isinstance(generator, (DirectedGraphSAGENodeGenerator,)):
```
Should this test include `DirectedGraphSAGELinkGenerator`?

adocherty replied:
It currently can't include `DirectedGraphSAGELinkGenerator` as we don't have one.  We should get onto that sometime.

geoffj-d61 commented on stellargraph/layer/graphsage.py:
```diff
@@ -1048,56 +1086,70 @@ class DirectedGraphSAGE(GraphSAGE):
-        ):
-            raise ValueError(
-                "If generator is not provided, in_samples, out_samples and input_dim must be specified."
+        try:
```
Interesting pattern. I know there are people who argue stridently against using try/except unless it is absolutely necessary. I guess I lean that way a bit too (hence the way I originally wrote the code), but I'm not fussed. At least your code does make it clear that those parameters are required.

adocherty replied:
Yes, I'm somewhat ambivalent about it myself.  I felt it made things neater here though.

youph commented:
Nice one!

youph commented on stellargraph/layer/graphsage.py:
```diff
@@ -808,14 +841,22 @@ def _get_sizes_from_generator(self, generator):
-        self.n_samples = generator.generator.num_samples
+        if not isinstance(generator, (GraphSAGELinkGenerator, GraphSAGENodeGenerator)):
+            raise TypeError(
+                "Generator should be an instance of GraphSAGELinkGenerator or GraphSAGENodeGenerator"
```
swap the order to "GraphSAGENodeGenerator or GraphSAGELinkGenerator"?

adocherty replied:
Seems pretty minor. Made the change.

youph commented on demos/link-prediction/attri2vec/stellargraph-attri2vec-DBLP.ipynb:
```diff
@@ -30,7 +30,38 @@
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "/Users/doc019/.envs/sgdev/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:516: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
```
could you please clean the notebooks from user-specific metadata?

adocherty replied:
I've cleaned the notebooks and created a new script that auto-cleans the notebooks for us by removing tensorflow warnings, stderr messages, renumbering the cells, and setting the kernelspec to by "Python 3".

The new script is here `scripts/format_notebooks.py`

youph replied:
@adocherty nice! Did you write that script yourself, or got it from somewhere? Does it need attributions and CSIRO copyright header?

youph replied:
why did you delete the format notebook script?

adocherty commented:
>     * At least one python script had't been updated to the new API: `demos/node-classification/graphsage/graphsage-cora-example.py:121`. I've fixed this, but there are possibly more. Please check.

Fixed this and caught another script that was not migrated.

>     * This is a breaking change. What are your thoughts on handling user code from previous versions gracefully? I think the least we could do is give an informative message, say, in `stellargraph/layer/graphsage.py:846` we could expand the error message by saying something like "Passing a Sequence is no longer supported. Generator should be an instance of GraphSAGENodeGenerator or GraphSAGELinkGenerator" - or something to that effect. We should also say this in the CHANGELOG.

I've added a message along the lines you suggest to DirectedGraphSAGE, GraphSAGE & HinSAGE and updated the changelog.  We could deprecate this more gracefully by putting a `DeprecationWarning` into the code and creating some hacks to make this work.  Do you think this is worthwhile?

youph commented:
Nuh, I think what you did with the messages is good, no need for deprecation warnings.
